### PR TITLE
8273277: C2: Move conditional negation into rc_predicate

### PR DESCRIPTION
**Pull request by neliasso**
Hi,

I need some feedback on this patch. This was reported from Tencent and found in internal testing about the same time. This patch is based on a a patch provided by Tencent.

In some very specific circumstances we need to negate the range checks that we create in PhaseIdealLoop::loop_predication_impl_helper. This is done in three places, but that method also calls insert_initial_skeleton_predicate where this isn't taken into account. 

To simplify things I have moved the negation logic into rc_predicate. This should prevent us from missing this check again.

I do have a concern that negating the condition of the rangecheck in the skeleton predicate, since the skeleton predicate will be used as a clone template, and some rangechecks optimizations seems to assume that range checks always have LT as the condidtion. On the other hand - it is a really uncommon scenario since we haven't failed here before.

Feedback appreciated.

Best regards,
Nils

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8273277](https://bugs.openjdk.java.net/browse/JDK-8273277): C2: Move conditional negation into rc_predicate

### Reviewers
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Christian Hagedorn](https://openjdk.java.net/census#chagedorn) (@chhagedorn - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5987/head:pull/5987` \
`$ git checkout pull/5987`

Update a local copy of the PR: \
`$ git checkout pull/5987` \
`$ git pull https://git.openjdk.java.net/jdk pull/5987/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5987`

View PR using the GUI difftool: \
`$ git pr show -t 5987`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5987.diff">https://git.openjdk.java.net/jdk/pull/5987.diff</a>

</details>

### CODE REVIEW

**Comment by neliasso** [comment]
> > I do have a concern that negating the condition of the rangecheck in the skeleton predicate, since the skeleton predicate will be used as a clone template, and some rangechecks optimizations seems to assume that range checks always have LT as the condition. On the other hand - it is a really uncommon scenario since we haven't failed here before.
> 
> Is there any specific code that you worry about?

I have seen checks for LT, but I see none that specifically would be affecting this change.

> 
> I think it should be fine because the purpose of copying and instantiating skeleton range check predicates is to guarantee that control/data paths die consistently when the main loop induction variable falls outside of the allowed range of an array access. But @rwestrel and @chhagedorn looked more into this recently.
> 
> Can we add the test that Tencent found as well?

Yes - I've added it.

> 
> Please update the affects versions in the bug report.

Done.

File: test/hotspot/jtreg/compiler/loopopts/TestSkeletonPredicateNegation.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/**
+ * @test
+ * @bug 8273277
+ * @summary Skeleton predicates sometimes need to be negated
+ * @run main compiler.loopopts.TestSkeletonPredicateNegation
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/loopopts/TestSkeletonPredicateNegation.java:29]: I think this should be changed to `@run driver`.

File: test/hotspot/jtreg/compiler/loopopts/TestSkeletonPredicateNegation.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                instance.mainTest(args);
+            }
+        } catch (Exception ex) {
+            System.out.println(ex.getClass().getCanonicalName());
+        } catch (OutOfMemoryError e) {
+            System.out.println("OOM Error");
+        }
+    }
+
+    public void mainTest (String[] args){
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/loopopts/TestSkeletonPredicateNegation.java:51]: Code style should be `mainTest(String[] args) {`

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Review by chhagedorn — APPROVED** [review]
That looks reasonable to me.

**Review by vnkozlov — APPROVED** [review]
I agree.

**Pull request closed without merging**